includes/class-wc-payments-utils.php

```diff
@@ -477,4 +479,31 @@ public static function convert_to_stripe_locale( string $locale ): string {
 		// Return 'auto' so Stripe.js uses the browser locale.
 		return 'auto';
 	}
+
+	/**
+	 * Returns redacted customer-facing error messages for notices.
+	 *
+	 * This function tries to filter out API exceptions that should not be displayed to customers.
+	 * Generally, only Stripe exceptions with type of `card_error` should be displayed.
+	 * Other API errors should be redacted (https://stripe.com/docs/api/errors#errors-message).
+	 *
+	 * @param Exception $e Exception to get the message from.
+	 *
+	 * @return string
+	 */
+	public static function get_filtered_error_message( Exception $e ) {
+		$error_message = method_exists( $e, 'getLocalizedMessage' ) ? $e->getLocalizedMessage() : $e->getMessage();
+
+		// These notices can be shown when placing an order or adding a new payment method, so we aim for
+		// more generic messages instead of specific order/payment messages when the API Exception is redacted.
+		if ( $e instanceof Connection_Exception ) {
+			$error_message = __( 'There was an error while processing this request. If you continue to see this notice, please contact the admin.', 'woocommerce-payments' );
+		} elseif ( $e instanceof API_Exception && 'wcpay_bad_request' === $e->get_error_code() ) {
```

Comment:
Do you think we should filter other (or all) WCPay server error codes too? The only example I have in mind is if the account (hypothetically) isn't found when calling `/intentions`.

Reply:
> Do you think we should filter other (or all) WCPay server error codes too?

@dechov I quickly looked into the `/intentions` request and it seems like it should throw an `API_Exception`, which should be filtered out by [this line](https://github.com/Automattic/woocommerce-payments/pull/2502/files#diff-fdf97bb0bd6043ee97d2ab4f058059c59d2b7173a076398fe8441824dc7b5e35R503).

But I see there are a few other API Exceptions that wouldn't be filtered out in `class-wc-payments-api-client.php`. Perhaps we could filter all API Exceptions and only display the `card_error` ones?

Reply:
That line only filters out exceptions that have a `type`, which (correct me if I'm wrong) is specific to Stripe's error responses and which we still aren't adding to the exception (i.e. leaving it `null`) in most cases.

I'd be fine with leaving it for a later PR or issue if you'd prefer, since the errors originating from WCPay server aren't specifically documented as customer-facing (if ever) or not, like the ones originating from Stripe are, so it might need more consideration.

Reply:
> That line only filters out exceptions that have a `type`

That's correct.

Not all WCPay server errors would be filtered out. For that I think we would need to scope those errors from the server, which would be easier to do from a server PR.

@dechov what about creating a follow up issue for #470 noting that we need to check all custom errors thrown by the server and then filter out the ones that are not supposed to be displayed to customers? This might require a server and another client PR.

Reply:
Sounds fine to me – I guess technically we could just add `type` to the server errors.

Reply:
1174-gh-Automattic/woocommerce-payments-server

I created a server issue as a spike to first check if there are any cases where we'd like to filter out error messages, and then propose a way to differentiate these errors. Adding a `type` is a good idea, I added to the issue.